File: go/definition/definition.go
package definition

const DEFINITION_VERSION_MAJOR = 1
const DEFINITION_VERSION_MINOR = 0
const DEFINITION_VERSION = "1.0"

type JPLDefinition struct {
	Version      string           `json:"version"`
	Instructions []JPLInstruction `json:"instructions"`
}

type Pipe = []JPLInstruction

type JPLInstruction struct {
	OP     JPLOP                `json:"op"`
	Params JPLInstructionParams `json:"params"`
}

type JPLInstructionParams struct {
	After          string             `json:"after,omitempty"`
	ArgNames       []string           `json:"argNames,omitempty"`
	Assignment     *JPLAssignment     `json:"assignment,omitempty"`
	Catch          Pipe               `json:"catch,omitempty"`
	Comparisons    []JPLComparison    `json:"comparisons,omitempty"`
	Else           Pipe               `json:"else,omitempty"`
	Fields         []JPLField         `json:"fields,omitempty"`
	Ifs            []JPLIfThen        `json:"ifs,omitempty"`
	Interpolations []JPLInterpolation `json:"interpolations,omitempty"`
	Name           string             `json:"name,omitempty"`
	Operations     []JPLOperation     `json:"operations,omitempty"`
	Pipe           Pipe               `json:"pipe,omitempty"`
	Pipes          []Pipe             `json:"params,omitempty"`
	Selectors      []JPLSelector      `json:"selectors,omitempty"`
	Try            Pipe               `json:"try,omitempty"`
	Value          any                `json:"value,omitempty"`
	String         string             `json:"string,omitempty"`
	Number         float64            `json:"number,omitempty"`
}

type JPLComparison struct {
	OP     JPLOPC              `json:"op"`
	Params JPLComparisonParams `json:"params"`
}

type JPLComparisonParams struct {
	By Pipe `json:"by"`
}

type JPLSelector struct {
	OP     JPLOPA            `json:"op"`
	Params JPLSelectorParams `json:"params"`
}

type JPLSelectorParams struct {
	Args     []Pipe `json:"args,omitempty"`
	Bound    bool   `json:"bound,omitempty"`
	From     Pipe   `json:"from,omitempty"`
	Optional bool   `json:"optional,omitempty"`
	Pipe     Pipe   `json:"pipe,omitempty"`
	To       Pipe   `json:"to,omitempty"`
}

type JPLOperation struct {
	OP     JPLOPM             `json:"op"`
	Params JPLOperationParams `json:"params"`
}

type JPLOperationParams struct {
	By Pipe `json:"by,omitempty"`
}

type JPLIfThen struct {
	If   Pipe `json:"if"`
	Then Pipe `json:"then"`
}

type JPLAssignment struct {
	OP     JPLOPU              `json:"op"`
	Params JPLAssignmentParams `json:"params"`
}

type JPLAssignmentParams struct {
	Pipe Pipe `json:"pipe"`
}

type JPLField struct {
	Key      Pipe `json:"key"`
	Value    Pipe `json:"value"`
	Optional bool `json:"optional,omitempty"`
}

type JPLInterpolation struct {
	Before string `json:"before"`
	Pipe   Pipe   `json:"pipe"`
}

File: js/lib/library/definition.ts
import {
  OPU_MULTIPLICATION,
  OPU_NULL_COALESCENCE,
  OPU_REMAINDER,
  OPU_SET,
  OPU_SUBTRACTION,
  OPU_UPDATE,
  type OP_ACCESS,
  type OP_AND,
  type OP_ARRAY_CONSTRUCTOR,
  type OP_ASSIGNMENT,
  type OP_CALCULATION,
  type OP_COMPARISON,
  type OP_CONSTANT,
  type OP_CONSTANT_FALSE,
  type OP_CONSTANT_NULL,
  type OP_CONSTANT_TRUE,
  type OP_FUNCTION_DEFINITION,
  type OP_IF,
  type OP_INTERPOLATED_STRING,
  type OP_NEGATION,
  type OP_NOT,
  type OP_NULL_COALESCENCE,
  type OP_NUMBER,
  type OP_OBJECT_CONSTRUCTOR,
  type OP_OR,
  type OP_OUTPUT_CONCAT,
  type OP_STRING,
  type OP_TRY,
  type OP_VARIABLE,
  type OP_VARIABLE_DEFINITION,
  type OP_VOID,
  type OPA_FIELD,
  type OPA_FUNCTION,
  type OPA_ITER,
  type OPA_SLICE,
  type OPC_EQUAL,
  type OPC_GREATER,
  type OPC_GREATEREQUAL,
  type OPC_LESS,
  type OPC_LESSEQUAL,
  type OPC_UNEQUAL,
  type OPM_ADDITION,
  type OPM_DIVISION,
  type OPM_MULTIPLICATION,
  type OPM_REMAINDER,
  type OPM_SUBTRACTION,
  type OPU_ADDITION,
  type OPU_DIVISION,
} from './ops';

export const DEFINITION_VERSION_MAJOR = 1;
export const DEFINITION_VERSION_MINOR = 0;
export const DEFINITION_VERSION = `${DEFINITION_VERSION_MAJOR}.${DEFINITION_VERSION_MINOR}`;

export type JPLDefinition = {
  version: string;
  instructions: Pipe;
};

export type Pipe = JPLInstruction[];

export type JPLInstruction = {
  op:
    | typeof OP_ACCESS
    | typeof OP_AND
    | typeof OP_ARRAY_CONSTRUCTOR
    | typeof OP_ASSIGNMENT
    | typeof OP_CALCULATION
    | typeof OP_COMPARISON
    | typeof OP_CONSTANT
    | typeof OP_CONSTANT_FALSE
    | typeof OP_CONSTANT_NULL
    | typeof OP_CONSTANT_TRUE
    | typeof OP_FUNCTION_DEFINITION
    | typeof OP_IF
    | typeof OP_INTERPOLATED_STRING
    | typeof OP_NEGATION
    | typeof OP_NOT
    | typeof OP_NULL_COALESCENCE
    | typeof OP_NUMBER
    | typeof OP_OBJECT_CONSTRUCTOR
    | typeof OP_OR
    | typeof OP_OUTPUT_CONCAT
    | typeof OP_STRING
    | typeof OP_TRY
    | typeof OP_VARIABLE
    | typeof OP_VARIABLE_DEFINITION
    | typeof OP_VOID;
  params: JPLInstructionParams;
};

export type JPLInstructionParams = {
  after?: string;
  argNames?: string[];
  assignment?: JPLAssignment;
  catch?: Pipe;
  comparisons?: JPLComparison[];
  else?: Pipe;
  fields?: JPLField[];
  ifs?: JPLIfThen[];
  interpolations?: JPLInterpolation[];
  name?: string;
  number?: number;
  operations?: JPLOperation[];
  pipe?: Pipe;
  pipes?: Pipe[];
  selectors?: JPLSelector[];
  string?: string;
  try?: Pipe;
  value?: unknown;
};

export type JPLComparison = {
  op:
    | typeof OPC_EQUAL
    | typeof OPC_GREATER
    | typeof OPC_GREATEREQUAL
    | typeof OPC_LESS
    | typeof OPC_LESSEQUAL
    | typeof OPC_UNEQUAL;
  params: JPLComparisonParams;
};

export type JPLComparisonParams = { by: Pipe };

export type JPLSelector = {
  op:
    | typeof OPA_FIELD
    | typeof OPA_FUNCTION
    | typeof OPA_ITER
    | typeof OPA_SLICE;
  params: JPLSelectorParams;
};

export type JPLSelectorParams = {
  args?: Pipe[];
  bound?: boolean;
  from?: Pipe;
  optional?: boolean;
  pipe?: Pipe;
  to?: Pipe;
};

export type JPLOperation = {
  op:
    | typeof OPM_ADDITION
    | typeof OPM_DIVISION
    | typeof OPM_MULTIPLICATION
    | typeof OPM_REMAINDER
    | typeof OPM_SUBTRACTION;
  params: JPLOperationParams;
};

export type JPLOperationParams = { by: Pipe };

export type JPLIfThen = { if: Pipe; then: Pipe };

export type JPLAssignment = {
  op:
    | typeof OPU_ADDITION
    | typeof OPU_DIVISION
    | typeof OPU_MULTIPLICATION
    | typeof OPU_NULL_COALESCENCE
    | typeof OPU_REMAINDER
    | typeof OPU_SET
    | typeof OPU_SUBTRACTION
    | typeof OPU_UPDATE;
  params: JPLAssignmentParams;
};

export type JPLAssignmentParams = { pipe: Pipe };

export type JPLField = { key: Pipe; value: Pipe; optional: boolean };

export type JPLInterpolation = { before: string; pipe: Pipe };
